File: answers/leetcode/Serialize and Deserialize Binary Tree/Serialize and Deserialize Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        if (root != null) {
            sb.append(root.val)
                .append(",")
                .append(this.serialize(root.left))
                .append(",")
                .append(this.serialize(root.right));
        } else {
            sb.append("null");
        }
        return sb.toString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] splitted = data.split(",");
        this.index = 0;
        return this.deserialize(splitted);
    }
    
    private int index;
    
    private TreeNode deserialize(String[] splitted) {
        if (splitted == null || splitted.length == 0 || splitted[this.index].equals("null")) {
            this.index++;
            return null;
        }
        TreeNode root = new TreeNode(Integer.parseInt(splitted[this.index]));
        this.index++;
        root.left = this.deserialize(splitted);
        root.right = this.deserialize(splitted);
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: answers/leetcode/Serialize and Deserialize Binary Tree/Serialize and Deserialize Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string serial;
        if (root) {
            serial = to_string(root->val) + string(",") + serialize(root->left) + string(",") + serialize(root->right);
        }
        else {
            serial = string("null");
        }
        return serial;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        size_t pos = 0;
        return deserialize1(pos, data);
    }
    TreeNode* deserialize1(size_t &current, const string &data) {
        TreeNode *root = NULL;
        if (current >= data.length()) {
            return root;
        }
        size_t delimiter = data.find(',', current);
        string root_str = data.substr(current, delimiter - current);
        current = delimiter;
        if (string("null") != root_str) {
            root = new TreeNode(atoi(root_str.c_str()));
            ++ current;
            root->left = deserialize1(current, data);
            ++ current;
            root->right = deserialize1(current, data);
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));